From 7fac345a5c2ccb48c7194565443ec9168a97a8cf Mon Sep 17 00:00:00 2001
From: Khalid Abu Shawarib <khalid.shawarib@gmail.com>
Date: Tue, 22 Oct 2024 16:15:29 +0300
Subject: [PATCH] xdp-gnome/externalwindow: Call gtk_init before opening a
 display

See https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/7836#note_2254328

Fixes: https://gitlab.gnome.org/GNOME/nautilus/-/issues/3627


(cherry picked from commit 8f4242b1ecdc12a7bc8900bfb6531af6b2584c6b)
---
 xdp-gnome/externalwindow-wayland.c | 16 ++++++++++++----
 xdp-gnome/externalwindow-x11.c     | 11 ++++++++++-
 2 files changed, 22 insertions(+), 5 deletions(-)

diff --git a/xdp-gnome/externalwindow-wayland.c b/xdp-gnome/externalwindow-wayland.c
index a9201396a0..4c3020cf01 100644
--- a/xdp-gnome/externalwindow-wayland.c
+++ b/xdp-gnome/externalwindow-wayland.c
@@ -292,11 +292,19 @@ init_external_window_wayland_display (GError **error)
   fd_str = g_strdup_printf ("%d", fd);
 
   g_setenv ("WAYLAND_SOCKET", fd_str, TRUE);
+
+  g_assert (!gtk_is_initialized ());
   gdk_set_allowed_backends ("wayland");
-  display = gdk_display_open (NULL);
-  g_assert (display);
+  if (gtk_init_check ())
+    {
+      display = gdk_display_get_default ();
+
+      g_assert (GDK_IS_WAYLAND_DISPLAY (display));
 
-  init_x11_interop (display);
+      init_x11_interop (display);
 
-  return display;
+      return display;
+    }
+  else
+    return NULL;
 }
diff --git a/xdp-gnome/externalwindow-x11.c b/xdp-gnome/externalwindow-x11.c
index 151922b4fd..c2392842ea 100644
--- a/xdp-gnome/externalwindow-x11.c
+++ b/xdp-gnome/externalwindow-x11.c
@@ -117,6 +117,15 @@ external_window_x11_class_init (ExternalWindowX11Class *klass)
 GdkDisplay *
 init_external_window_x11_display (GError **error)
 {
+  g_assert (!gtk_is_initialized ());
   gdk_set_allowed_backends ("x11");
-  return gdk_display_open (NULL);
+  if (gtk_init_check ())
+    {
+      GdkDisplay *display = gdk_display_get_default ();
+
+      g_assert (GDK_IS_X11_DISPLAY (display));
+      return display;
+    }
+  else
+    return NULL;
 }
-- 
GitLab

